src/Canvas.cc

```diff
@@ -879,18 +879,21 @@ Canvas::ResolveFontDescription(const PangoFontDescription *desc) {
       if (streq_casein(family, pangofamily)) {
         const char* sys_desc_family_name = pango_font_description_get_family(ff.sys_desc);
         bool unseen = seen_families.find(sys_desc_family_name) == seen_families.end();
+        bool better = best.user_desc == nullptr || pango_font_description_better_match(desc, best.user_desc, ff.user_desc);
 
         // Avoid sending duplicate SFNT font names due to a bug in Pango for macOS:
         // https://bugzilla.gnome.org/show_bug.cgi?id=762873
         if (unseen) {
           seen_families.insert(sys_desc_family_name);
-          if (renamed_families.size()) renamed_families += ',';
-          renamed_families += sys_desc_family_name;
-        }
 
-        if (first && (best.user_desc == nullptr || pango_font_description_better_match(desc, best.user_desc, ff.user_desc))) {
-          best = ff;
+          if (better) {
+            renamed_families = string(sys_desc_family_name) + (renamed_families.size() ? "," : "") + renamed_families;
+          } else {
+            renamed_families = renamed_families + (renamed_families.size() ? "," : "") + sys_desc_family_name;
+          }
         }
+
+        if (first && better) best = ff;
```

Comment:
Not tested, but could you get rid of `first` and replace this line with `if (best.user_desc == nullptr)`?

Reply:
That would be a behavior change since we weren't setting the `best` if the first family wasn't found in the registered fonts. If you specify `"Arimo", "Open Sans"` but you never registered `"Arimo"`, then `"Arimo", sys_desc("Open Sans")` goes to the OS with the rest of the user description untouched.

`renamed_families` is also not sorted correctly, it only guarantees that the top match is first rather than ranking them all, but it's better than before and I don't think many people even use the font cascade that way. This probably needs an overhaul. It's all so complicated when we don't have entirely internal font selection 😞